src/history/file_backed.rs

```diff
@@ -42,54 +45,90 @@ impl Default for FileBackedHistory {
     }
 }
 
-fn encode_entry(s: &str) -> String {
-    s.replace('\n', NEWLINE_ESCAPE)
+static ID_SEP: &str = "<id>:";
+
+fn encode_entry(id: HistoryItemId, s: &str) -> String {
+    format!("{id}{ID_SEP}{}", s.replace('\n', NEWLINE_ESCAPE))
 }
 
-fn decode_entry(s: &str) -> String {
-    s.replace(NEWLINE_ESCAPE, "\n")
+/// Decode an entry
+///
+/// Legacy format: ls /
+/// New format   : 182535<id>:ls /
+///
+/// If a line can't be parsed using the new format, it will fallback to the legacy one.
+///
+/// This allows this function to support decoding for both legacy and new histories,
+/// as well as mixing both of them.
+fn decode_entry(s: &str, counter: &mut i64) -> (HistoryItemId, String) {
+    let mut parsed = None;
+
+    if let Some(sep) = s.find(ID_SEP) {
+        if let Ok(parsed_id) = s[..sep].parse() {
+            parsed = Some((parsed_id, &s[sep + ID_SEP.len()..]));
+        }
+    }
+
+    let (id, content) = parsed.unwrap_or_else(|| {
+        *counter += 1;
+        (*counter - 1, s)
+    });
+
+    (HistoryItemId(id), content.replace(NEWLINE_ESCAPE, "\n"))
 }
 
 impl History for FileBackedHistory {
+    fn generate_id(&mut self) -> HistoryItemId {
+        HistoryItemId(self.rng.gen())
+    }
+
     /// only saves a value if it's different than the last value
-    fn save(&mut self, h: HistoryItem) -> Result<HistoryItem> {
-        let entry = h.command_line;
+    fn save(&mut self, h: &HistoryItem) -> Result<()> {
+        let entry = h.command_line.clone();
+
         // Don't append if the preceding value is identical or the string empty
-        let entry_id = if self
+        if self
             .entries
-            .back()
-            .map_or(true, |previous| previous != &entry)
+            .last()
+            .map_or(true, |(_, previous)| previous != &entry)
             && !entry.is_empty()
         {
-            if self.entries.len() == self.capacity {
+            if self.entries.len() >= self.capacity {
                 // History is "full", so we delete the oldest entry first,
                 // before adding a new one.
-                self.entries.pop_front();
-                self.len_on_disk = self.len_on_disk.saturating_sub(1);
+                let first_id = *(self.entries.first().unwrap().0);
```

Comment:
I'm not objecting to getting the first ID somewhere. I'm objecting to using unwrap. I'd guess that you can match on it and handle errors and use map_err if the error is a different flavor.

Reply:
This would be misleading. Using a `.map_err()` in a situation we've _proven_ to meet the condition required for unwrapping may lead other developers to think there is a case where this could fail - while it can't, that's why I used .unwrap() (which I always use very sparingly).

Writing `.first().unwrap()` is more readable, less confusing, and more correct as if we don't meet this condition something's **_very_** wrong and we cannot continue processing anything because some of the code is completely invalid.